### PR TITLE
bsc#1132881, add sharutils to package requirement for uuencode.

### PR DESCRIPTION
**Pull request by nick-wang**
Add requirement of package `sharutils` for binary `uuencode`.

### CODE REVIEW

**Comment by nick-wang** [comment]
The mess log is a bad consequence of using cherry-pick instead of merge in a previous commit...

File: src/modules/Cluster.rb [hunk 1]
```diff
@@ -628,26 +629,12 @@ def Write
...
-          SCR.Execute(
-            path(".target.bash_output"),
-            Ops.add(
-              Ops.add("echo '", @corokey),
-              "' | uudecode -o /etc/corosync/authkey"
-            )
-          )
-        )
+        file = File.open("/etc/corosync/authkey", "w")
+        file.write(Base64.decode64(@corokey))
```
Comment (lslezak) [inline src/modules/Cluster.rb]: You can replace these two lines by a single line:

```ruby
File.write("/etc/corosync/authkey", Base64.decode64(@corokey))
```

File: src/modules/Cluster.rb [hunk 2]
```diff
@@ -628,26 +629,12 @@ def Write
...
-          SCR.Execute(
-            path(".target.bash_output"),
-            Ops.add(
-              Ops.add(Ops.add("echo '", @csync2key), "' | uudecode -o "),
-              @csync2_key_file
-            )
-          )
-        )
+        file = File.open(@csync2_key_file, "w")
+        file.write(Base64.decode64(@csync2key))
```
Comment (lslezak) [inline src/modules/Cluster.rb]: The same here.

File: src/modules/Cluster.rb [hunk 3]
```diff
@@ -722,22 +709,12 @@ def Export
...
       end
-      if SCR.Read(path(".target.size"), "/etc/csync2/key_hagroup") != -1
-        out = Convert.to_map(
-          SCR.Execute(
-            path(".target.bash_output"),
-            Ops.add(Ops.add("uuencode -m ", @csync2_key_file), " /dev/stdout ")
-          )
-        )
-        Ops.set(result, "csync2key", Ops.get_string(out, "stdout", ""))
+      if SCR.Read(path(".target.size"), @csync2_key_file) != -1
```
Comment (lslezak) [inline src/modules/Cluster.rb]: You can use the native `File.exist?` here:
```suggestion
      if File.exist?(@csync2_key_file)
```

**Comment by nick-wang** [comment]
@lslezak Thanks for the review.

> I have just few small suggestions to the code, but there is a bigger problem:
> 
> Changing the file encoding from UU encoding to Base64 will make problems at upgrade. If you save the `/etc/corosync/authkey` file using the UU encoding then upgrade the package and then try to load it using the Base64 encoding then it will fail.
Indeed, it may cause an incompatible of upgrade. If encode with the old uuencode and decode with Base64. However, it should be fine as long as all nodes in cluster using the same way and the decode string doesn't include any invalid string.(I believe so for the later part)

However, even some of the nodes in cluster is not upgraded, it may need regenerate the authkey or manually sync key from the old node. Not so critical error. (A solution in the later part may help)

> 
> There is also another question: do some external tools (out of YaST) read that file? Changing the encoding might also some side effects there...
>
The encoding string is used for autoyast only, so i think no problem in this case. The external tools only read the decoded file like the authkey.
 
> Maybe we should add some backward compatibility, e.g. try to detect the file encoding. If the file only contains `[A-Z]` , `[a-z]`, `[0-9]` and `=+/` characters we could assume Base64, otherwise assume UU encoding. (And always save using Base64.)
Both uuencode and Base64 will encode as base64 format. However, the strings is different in my test environment IIRC.

> 
> Or because decoding an invalid string results in empty string (`` Base64.decode64("&*``#$^*") => "" ``) we could fallback to UU encoding if the decode result is empty string.
Since both of the method will encode as base4, so i think all strings are valid.

I have a solution for this part, like:
Check whether binary(uuencode/uudecode) or package(sharutils) installed. If so, using the prervious method. If not, using the Base64 library instead.
In this case, we could drop the `require: sharutils` in .spec file. Cause upgrade yast2-cluster won't trigger the deinstallation of `sharutils`. Nearly 0 chance that encode via `uuencode` and decode via `Base64`.

What is your opinion about this solution?

**Comment by nick-wang** [comment]
@lslezak Could you please help to review the new fix? This should compatible with uuencode/uudecode without add the require to sharutils.

**Comment by lslezak** [comment]
> I have a solution for this part, like:
> Check whether binary(uuencode/uudecode) or package(sharutils) installed. If so, using the prervious  method. If not, using the Base64 library instead.
>In this case, we could drop the require: sharutils in .spec file.

Um, the problem is that the `sharutils` package might be installed or uninstalled by some other package dependencies (or explicitly by the user).

So it might happen something like this:

```console
# Upgrade the yast2-cluster to use uudecode or Base64 depending whether
# the sharutils is installed. Assume, that sharutils is not installed yet (in
# reality the customers would rather run `zypper patch` but that does not matter,
# the result would be the same).
zypper up yast2-cluster

# now use the package, it writes the config in the Base64 encoding
yast2 cluster

# now install the sharutils package (or install any other package requiring it)
zypper install sharutils

# now use the package, it will read the Base64 config but do the UU decoding !
yast2 cluster
```

So I'd rather prefer some Base64/UU encoding detection as written above. Or alternatively you might write the used encoding somewhere. Either to a separate file, or directly into the file. 

E.g. write `encoding=base64` or `encoding=uu` as the very first line in the file. When reading use that specified encoding, if not found use the UU encoding as a fallback for the backward compatibility.

**Comment by nick-wang** [comment]
@lslezak Thanks a lot for the review and great help.

**Pull request merged**